server/hotrod/src/main/java/org/infinispan/server/hotrod/tx/TxState.java

```diff
@@ -0,0 +1,183 @@
+package org.infinispan.server.hotrod.tx;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+import org.infinispan.commands.write.WriteCommand;
+import org.infinispan.commons.marshall.AdvancedExternalizer;
+import org.infinispan.commons.marshall.MarshallUtil;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.server.core.ExternalizerIds;
+import org.infinispan.transaction.xa.GlobalTransaction;
+
+import net.jcip.annotations.Immutable;
+
+/**
+ * A transaction state stored globally in all the cluster members.
+ *
+ * @author Pedro Ruivo
+ * @since 9.1
+ */
+@Immutable
+public class TxState {
+
+   public static final AdvancedExternalizer<TxState> EXTERNALIZER = new Externalizer();
+   private static final TxStatus[] CACHED_VALUE = TxStatus.values();
+   private final GlobalTransaction globalTransaction;
+   private final TxStatus status;
+   private final List<WriteCommand> modifications;
+
+   TxState(GlobalTransaction globalTransaction) {
+      this(globalTransaction, TxStatus.RUNNING, null);
+   }
+
+   private TxState(GlobalTransaction globalTransaction, TxStatus status, List<WriteCommand> modifications) {
+      this.globalTransaction = globalTransaction;
+      this.status = status;
+      this.modifications = modifications;
+   }
+
+   private static TxStatus valueOf(int i) {
+      return CACHED_VALUE[i];
+   }
+
+   private static List<WriteCommand> copyModifications(Collection<WriteCommand> modifications) {
+      return modifications == null || modifications.isEmpty() ?
+            null :
+            Collections.unmodifiableList(new ArrayList<>(modifications));
+   }
+
+   private static List<WriteCommand> unmarshallModifications(ObjectInput input)
+         throws IOException, ClassNotFoundException {
+      List<WriteCommand> modifications = MarshallUtil.unmarshallCollection(input, ArrayList::new);
+      return modifications == null ? null : Collections.unmodifiableList(modifications);
```

Comment:
do we need to be *that* defensive when the `getModifications` is package-private?